qlib/workflow/__init__.py

```diff
@@ -0,0 +1,157 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+from contextlib import contextmanager
+from .record import MLflowRecorder
+from .exp import MLflowExpManager
+
+class Record:
+    def __init__(self):
+        pass
+
+    @contextmanager
+    def start_exp(self, experiment_name=None, uri=None, project_path=None, artifact_location=None, nested=False):
+        raise NotImplementedError(f"Please implement the `start_exp` method.")
+    
+    def search_runs(self, experiment_ids=None, filter_string='', run_view_type=1, max_results=100000, order_by=None):
+        raise NotImplementedError(f"Please implement the `search_runs` method.")
+    
+    def get_exp(self, experiment_id):
+        raise NotImplementedError(f"Please implement the `get_exp` method.")
+
+    def get_exp_by_name(self, experiment_name):
+        raise NotImplementedError(f"Please implement the `get_exp_by_name` method.")
+    
+    def create_exp(self, experiment_name, artifact_location=None):
+        raise NotImplementedError(f"Please implement the `create_exp` method.")
+    
+    def set_exp(self, experiment_name):
+        raise NotImplementedError(f"Please implement the `set_exp` method.")
+
+    def delete_exp(self, experiment_id):
+        raise NotImplementedError(f"Please implement the `create_exp` method.")
+
+    def set_tracking_uri(self, uri):
+        raise NotImplementedError(f"Please implement the `set_tracking_uri` method.")
+    
+    def get_tracking_uri(self):
+        raise NotImplementedError(f"Please implement the `get_tracking_uri` method.")
+    
+    def get_recorder(self):
+        raise NotImplementedError(f"Please implement the `get_recorder` method.")
+
+    def save_object(self, name, data):
+        raise NotImplementedError(f"Please implement the `save_object` method.")
+
+    def save_objects(self, name_data_list):
+        raise NotImplementedError(f"Please implement the `save_objects` method.")
+
+    def load_object(self, name):
+        raise NotImplementedError(f"Please implement the `load_object` method.")
+    
+    def log_param(self, key, value):
```

Comment:
I think we can merge  `log_param` and `log_params` by Python's  `**kwargs`  mechanism

Reply:
This trick applies to other functions